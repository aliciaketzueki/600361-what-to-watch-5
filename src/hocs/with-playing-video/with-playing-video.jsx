import React, {PureComponent, createRef} from "react";

const withPlayingVideo = (Component) => {
  class WithPlayingVideo extends PureComponent {
    constructor(props) {
      super(props);

      this.state = {
        isPlaying: false,
        duration: 0,
        progress: 0
      };

      this.onPlay = this.onPlay.bind(this);
      this.onPause = this.onPause.bind(this);
      this.onMouseDown = this.onMouseDown.bind(this);
      this.onFullScreen = this.onFullScreen.bind(this);
      this.video = createRef();
      this.progressBar = createRef();
      this.interval = 0;
    }

    componentDidMount() {
      const video = this.video.current;

      // if (!video) {
      //   return false;
      // }

      video.onstarted = () => {
        this.setState({
          duration: video.duration,
          isPlaying: true,
          progress: 0,
        });

        video.play();
      };

      video.onended = () => {
        this.setState({
          duration: 0,
          isPlaying: false,
        });

        video.pause();
      };

      video.ontimeupdate = () => {
        this.setState({
          duration: video.duration - video.currentTime,
          progress: video.currentTime / video.duration * 100
        });
      };
    }

    componentWillUnmount() {
      clearInterval(this.interval);
    }

    onPlay(evt) {
      evt.preventDefault();
      this.setState({isPlaying: true});
      this.video.current.play();
    }

    onPause(evt) {
      evt.preventDefault();
      this.setState({isPlaying: false});
      this.video.current.pause();
    }

    onMouseDown(evt) {
      let target = evt.target;
      let startCoords = evt.clientX;
      const progressBarWidth = this.progressBar.current.offsetWidth;

      document.onmousemove = (moveEvt) => {
        let shift = startCoords - moveEvt.clientX;
        startCoords = moveEvt.clientX;
        let left = target.offsetLeft - shift;
        let percentCoords = left * 100 / progressBarWidth;
        let timeByCoords = left * this.video.current.duration / progressBarWidth;

        if (percentCoords < 100 && percentCoords > 0) {
          this.video.current.currentTime = timeByCoords;

          this.setState({
            duration: this.video.current.duration - this.video.current.currentTime,
            progress: percentCoords
          });
        }
      };

      document.onmouseup = () => {
        document.onmousemove = null;
        document.onmouseup = null;
      };

      document.ondragstart = () => {
        return false;
      };
    }

    onFullScreen() {
      this.video.current.requestFullscreen();
    }

    render() {
      const {isPlaying, duration, progress} = this.state;

      return (
        <Component
          {...this.props}
          isPlaying={isPlaying}
          duration={duration}
          progress={progress}
          video={this.video}
          progressBar={this.progressBar}
          onPlay={this.onPlay}
          onPause={this.onPause}
          onMouseDown={this.onMouseDown}
          onFullScreen={this.onFullScreen}
        />
      );
    }
  }

  return WithPlayingVideo;
};

export default withPlayingVideo;
