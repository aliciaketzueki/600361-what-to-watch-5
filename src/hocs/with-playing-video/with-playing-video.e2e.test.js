import React from "react";
import {configure, mount} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import withPlayingVideo from "./with-playing-video";
import {film, noop, MockComponent} from "./with-playing-video.test";

configure({adapter: new Adapter()});

const MockComponentWrapped = withPlayingVideo(MockComponent);

describe(`Change state with video playing`, () => {
  it(`Should state changing depending on video playing`, () => {
    const wrapper = mount(
        <MockComponentWrapped
          film={film}
          isPlaying={false}
          onPlay={noop}
          onPause={noop}
          onMouseDown={noop}
          onFullScreen={noop}
          onExitBtnClick={noop}
        />, {
          createNodeMock() {
            return {};
          }
        }
    );

    expect(wrapper.state().isPlaying).toEqual(false);

    const e = {preventDefault: jest.fn()};
    window.HTMLMediaElement.prototype.play = noop;
    window.HTMLMediaElement.prototype.pause = noop;

    wrapper.instance().onPlay(e);
    expect(wrapper.state().isPlaying).toEqual(true);

    wrapper.instance().onPause(e);
    expect(wrapper.state().isPlaying).toEqual(false);
  });
});
