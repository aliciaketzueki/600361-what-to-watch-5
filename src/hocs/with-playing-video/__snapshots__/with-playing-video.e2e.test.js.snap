// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render withPlayingVideo Should withPlayingVideo render correctly 1`] = `
<div>
  <video
    className="player__video"
    poster="img/the-grand-budapest-hotel-bg.jpg"
    preload="metadata"
    src="https://some-link"
  />
</div>
`;
